### PR TITLE
Update to segmentation comparison

### PR DESCRIPTION
**Pull request by EStorvik**
Created a class structure for the segmentation comparison to help in visualizing the compared images. It should not in any way affect other parts of DarIA, so we could also discuss if it actually should not be a part of DarIA at all. Anyways, I'm creating this PR, and we can have a small discussion. No need to look carefully into the details (although they should not be too complicated anywhere).

### CODE REVIEW

File: src/daria/analysis/segmentationcomparison.py [hunk 1]
```diff
@@ -1,113 +1,471 @@
...
+    Routines for comparing segmentations and creating visualizations of the comparison.
+
+    Attributes:
+        number_of_segmented_images (int): Number of segmented images that one compares
+        segmentation_names (list[str]): list of names for each of the segmentations.
+            Will affect legends in plots.
+        components (list): list of values the different (active) components in the
+            segmentations. As of now, up to two are allowed for and the default values
+            are 1 and 2.
+        non_active_component (int): value for the nonactive component in the
```
Comment (jwboth) [inline src/daria/analysis/segmentationcomparison.py:28]: why is this needed, when `components` needs to be provided? Are not all other components inactive?

File: src/daria/analysis/segmentationcomparison.py [hunk 2]
```diff
@@ -1,113 +1,471 @@
...
+        self.component_names: list = kwargs.pop(
+            "component_names", ["Component 0", "Component 1"]
+        )
+
+        # Define gray colors
+        self.gray_colors: np.ndarray = kwargs.pop(
+            "gray_colors",
+            np.array(
+                [[180, 180, 180], [220, 220, 220], [200, 200, 200]], dtype=np.uint8
+            ),
```
Comment (jwboth) [inline src/daria/analysis/segmentationcomparison.py:86]: If really only gray colors are intended, wouldn't it be better to only ask for monochromatic values instead of trichromatic?

File: src/daria/analysis/segmentationcomparison.py [hunk 3]
```diff
@@ -1,113 +1,471 @@
...
+                f"{i+2} segmentations overlaps in {self.component_names[0]}"
+            ] = self.gray_base[0] * (i + 1)
+            self.color_dictionary[
+                f"{i+2} segmentations overlaps in {self.component_names[1]}"
+            ] = self.gray_base[1] * (i + 1)
+            self.color_dictionary[
+                f"Mixture {self.component_names[0]} and "
+                f"{self.component_names[1]} for {i+2} segmentations"
+            ] = self.gray_base[2] * (i + 1)
+
```
Comment (jwboth) [inline src/daria/analysis/segmentationcomparison.py:141]: This only allows for two consecutive images to overlap. When considering however three or more images, there is also other combinations than 0 and 1, as well as 1 and 2. A simple nested for loop structure should solve the problem.

File: src/daria/analysis/segmentationcomparison.py [hunk 4]
```diff
@@ -1,113 +1,471 @@
...
+        """
+        Comparison of segmentations.
+
+        Args:
+            segmentations (asterisk argument): Allows to provide an arbitraty
+                number of segmented numpy arrays or da.Images of integers to be compared
+            Optional keyword arguments (kwargs):
+                plot_result (bool): plots the result with matplotlib if True,
+                    default is False.
+                roi (tuple): roi where the segmentations should be compared,
```
[the remark this answers is not in the document]
Comment (jwboth) [reply src/daria/analysis/segmentationcomparison.py:157]: Also mention that the ROI should be provdide using matrix indexing.

[the remark this answers is not in the document]
Comment (jwboth) [reply src/daria/analysis/segmentationcomparison.py:157]: and addresses pixel coordinates (not metric units).

File: src/daria/analysis/segmentationcomparison.py [hunk 5]
```diff
@@ -1,113 +1,471 @@
...
+                dtype=np.uint8,
+            )
+
+        else:
+            if isinstance(segmentations[0], np.ndarray):
+                nx = min([seg.shape[0] for seg in segmentations])
+                ny = min([seg.shape[1] for seg in segmentations])
+            elif isinstance(segmentations[0], da.Image):
+                nx = min([seg.img.shape[0] for seg in segmentations])
+                ny = min([seg.img.shape[1] for seg in segmentations])
```
Comment (jwboth) [inline src/daria/analysis/segmentationcomparison.py:179]: This is confusing. nx and ny should be reversed. I woudl rather use `rows` and `cols` to reduce confusion.

File: src/daria/analysis/segmentationcomparison.py [hunk 6]
```diff
@@ -1,113 +1,471 @@
...
+            roi = (slice(0, nx), slice(0, ny))
+            return_image: np.ndarray = np.zeros((nx, ny) + (3,), dtype=np.uint8)
+
+        # Enter gray everywhere there are ovelaps of different segmentations
+
+        if isinstance(segmentations[0], np.ndarray):
```
Comment (jwboth) [inline src/daria/analysis/segmentationcomparison.py:185]: The `if` and `else` statements are exactly the same up to the way of addressing the arrays. Making use of call by reference, I guess, it would be cleaner to first set up references to the corresponding objects depending on the data type and then have a shorter (and single) analysis.

File: src/daria/analysis/segmentationcomparison.py [hunk 7]
```diff
@@ -1,113 +1,471 @@
...
+                    return_image[
+                        np.logical_or(
+                            np.logical_and(
+                                segmentations[k][roi] == self.components[0],
+                                segmentations[i][roi] == self.components[1],
+                            ),
+                            np.logical_and(
+                                segmentations[k][roi] == self.components[1],
+                                segmentations[i][roi] == self.components[0],
+                            ),
```
Comment (jwboth) [inline src/daria/analysis/segmentationcomparison.py:208]: To prepare for multiple component, it would be simpler to simple check whether segmentations[k][roi] values lie in self.components and segmentations[k][roi] is not equal segmentation[i][roi]. This shoudl be 2-3 lines and generalize.

File: src/daria/analysis/segmentationcomparison.py [hunk 8]
```diff
@@ -1,113 +1,471 @@
...
+                        )
+                    ] += self.gray_base[2]
+
+            # Determine locations (and make modifications to return image) of unique components
+            for c_num, c in enumerate(self.components):
+                for k in range(self.number_of_segmented_images):
+                    only_tmp: np.ndarray = segmentations[k][roi] == c
```
Comment (jwboth) [inline src/daria/analysis/segmentationcomparison.py:215]: I would try to use a more descriptive name and aim at not using anything with `tmp`.

File: src/daria/analysis/segmentationcomparison.py [hunk 9]
```diff
@@ -1,113 +1,471 @@
...
+                            np.logical_and(
+                                segmentations[k][roi] == self.components[0],
+                                segmentations[i][roi] == self.components[1],
+                            ),
+                            np.logical_and(
+                                segmentations[k][roi] == self.components[1],
+                                segmentations[i][roi] == self.components[0],
+                            ),
+                        )
+                    ] += self.gray_base[2]
```
Comment (jwboth) [inline src/daria/analysis/segmentationcomparison.py:210]: To be consistent with the last assignment
```suggestion
                    ] = self.gray_base[2]
```

File: src/daria/analysis/segmentationcomparison.py [hunk 10]
```diff
@@ -1,113 +1,471 @@
...
+                for i in range(k + 1, self.number_of_segmented_images):
+
+                    # Overlap of components
+                    for c_num, c in enumerate(self.components):
+                        return_image[
+                            np.logical_and(
+                                segmentations[k][roi] == c,
+                                segmentations[i][roi] == c,
+                            )
+                        ] += self.gray_base[c_num]
```
Comment (jwboth) [inline src/daria/analysis/segmentationcomparison.py:196]:
```suggestion
                        ] = self.gray_base[c_num]
```
To be consistent with the last assignment.

File: src/daria/analysis/segmentationcomparison.py [hunk 11]
```diff
@@ -1,113 +1,471 @@
...
+                    for j in filter(
+                        lambda j: j != k, range(self.number_of_segmented_images)
+                    ):
+                        only_tmp = np.logical_and(
+                            only_tmp,
+                            segmentations.img[j][roi] == self.non_active_component,
+                        )
 
-            return_image[only_tmp] = colors[k, c_num]
+                    return_image[only_tmp] = self.colors[k, c_num]
```
Comment (jwboth) [inline src/daria/analysis/segmentationcomparison.py:264]: I would try to remove this. This should be fixed by one line of code.

File: src/daria/analysis/segmentationcomparison.py [hunk 12]
```diff
@@ -1,113 +1,471 @@
...
+        # Get unique colors and sort them
+        unique_colors = self._get_unique_colors(comparison_image)
+        unique_colors = self._sort_colors(unique_colors)
+
+        # Process the comparison image
+        processed_comparison_image = self._post_process_image(
+            comparison_image,
+            unique_colors=unique_colors,
+            opacity=opacity,
+            countor_thickness=15,
```
Comment (jwboth) [inline src/daria/analysis/segmentationcomparison.py:460]:
```suggestion
            contour_thickness=15,
```

**Review by jwboth — APPROVED** [review]
I think it looks OK. I think one could improve the readibility by some more comments, and simplify the code here and there. But tit seems OK. Did you try the code with more than two images?

File: src/daria/analysis/segmentationcomparison.py [hunk 13]
```diff
@@ -1,113 +1,471 @@
...
+import matplotlib.pyplot as plt
 import numpy as np
+import skimage
 from matplotlib.cm import get_cmap
 
+import daria as da
+
 
-def compare_segmentations(*segmentations, **kwargs) -> np.ndarray:
+class CompareSegmentations:
```
[the remark this answers is not in the document]
Comment (EStorvik) [reply src/daria/analysis/segmentationcomparison.py:15]: Haha, I found it extremely much simpler in the end. Of course you were right all along! I totally agree on the name change. I don't know why I haven't changed it myself yet... at least three times I mistakenly called it using that name when I played around with it..

File: src/daria/analysis/segmentationcomparison.py [hunk 14]
```diff
@@ -1,113 +1,471 @@
...
+    def _post_process_image(
+        self,
+        image: np.ndarray,
+        unique_colors: Optional[np.ndarray] = None,
+        opacity: float = 1.0,
+        countor_thickness: int = 10,
+    ) -> np.ndarray:
+        """
+        Routine for post processing the image, removing background,
+        drawing countors and applying opacity.
```
[the remark this answers is not in the document]
Comment (EStorvik) [reply src/daria/analysis/segmentationcomparison.py:377]: Hahaha, clearly I have no clue how to spell contour...

File: src/daria/analysis/segmentationcomparison.py (hunk 1 above)
Comment (EStorvik) [reply src/daria/analysis/segmentationcomparison.py:28]: Sounds very correct, ill remove it!

File: src/daria/analysis/segmentationcomparison.py (hunk 2 above)
Comment (EStorvik) [reply src/daria/analysis/segmentationcomparison.py:86]: I thought about it, but decided that probably it could be a good idea to have the option to have other colors than gray in the future when more segmentations are compared. Moreover, I am not sure how easy it is to combine monochromatic and rgb colors in the same image?

File: src/daria/analysis/segmentationcomparison.py (hunk 3 above)
Comment (EStorvik) [reply src/daria/analysis/segmentationcomparison.py:141]: I don't understand what you mean? This is just the setup of the colors, it doesn't actually check which images overlap. Nevertheless, if we really want to distinguish the different gray colors for more than two images I do not think that the code can be kept like this, and we probably need to just choose specific color values for the cases we want to highlight.

File: src/daria/analysis/segmentationcomparison.py (hunk 6 above)
Comment (EStorvik) [reply src/daria/analysis/segmentationcomparison.py:185]: Good call!

File: src/daria/analysis/segmentationcomparison.py (hunk 7 above)
Comment (EStorvik) [reply src/daria/analysis/segmentationcomparison.py:208]: I made a suggestion now. It felt a bit more scary as it will return true if segmentations[i][roi] is a non active component where segmentations[k][roi] is in self.components, but that case is overwritten when checking for unique apperances.

File: src/daria/analysis/segmentationcomparison.py (hunk 10 above)
Comment (EStorvik) [reply src/daria/analysis/segmentationcomparison.py:196]: same?

**Comment by EStorvik** [comment]
I made several modifications according to your suggestions, and one very minor one in daria.Image. I was a bit afraid to put it here, but I just moved the toRGB() line, and thought it was ok to not make an entirely new pr for it. As it was implemented it didn't allow for monochromatic images to initilize the daria image, now it should be fine. If you disagree, ill remove the change from here, and make a separate PR. let me know.

**Review by jwboth — APPROVED** [review]
Thanks for simplifying the code and additing comments. I think it looks much better now!

**Comment by EStorvik** [comment]
I do too, thanks a lot!

File: src/daria/analysis/segmentationcomparison.py (hunk 9 above)
Comment (EStorvik) [reply src/daria/analysis/segmentationcomparison.py]: I am not sure what you mean?

**Comment by EStorvik** [comment]
Do you want to check over the unresolved conversations before I merge?

File: src/daria/analysis/segmentationcomparison.py (hunk 9 above)
Comment (jwboth) [reply src/daria/analysis/segmentationcomparison.py]: I did not get why you use summation here? after all, it seemed you treat disjoint cases anyhow. Overwriting seemed most natural. If I got something wrong, simply ignore my comment.

File: src/daria/analysis/segmentationcomparison.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jwboth) [reply src/daria/analysis/segmentationcomparison.py]: Looked OK, when I read it.

File: src/daria/analysis/segmentationcomparison.py (hunk 9 above)
Comment (EStorvik) [reply src/daria/analysis/segmentationcomparison.py]: My "secret" plan here, is that with the summation I can check how many are overlapping, and not just whether some are overlapping or not. I am almost sure that it will be too much information in the end (with 5 compared images), but I will keep it for now, and probably revert to the equality at a later point.

Comment (EStorvik) [reply src/daria/analysis/segmentationcomparison.py]: I tested with 4 images now, and it became a true mess. I will make the equality instead.

File: src/daria/analysis/segmentationcomparison.py (hunk 3 above)
Comment (EStorvik) [reply src/daria/analysis/segmentationcomparison.py]: This was indeed a mess, and I fixed it now.

Comment (jwboth) [reply src/daria/analysis/segmentationcomparison.py]: ah OK, I misunderstood the code I think. Sorry. I guess a comment could have helped, but it is fine.

File: src/daria/analysis/segmentationcomparison.py (hunk 9 above)
Comment (jwboth) [reply src/daria/analysis/segmentationcomparison.py]: Ah OK. In principle I am fine with these type of hacks. But then, good documentation is needed - for the future hacks.

File: src/daria/analysis/segmentationcomparison.py [hunk 15]
```diff
@@ -1,113 +1,442 @@
...
+        # Determine whether segmentations are arrays of daria images.
+        # They should all be the same.
+        if all([isinstance(seg, np.ndarray) for seg in segmentations]):
+            segmentation_arrays: tuple[np.ndarray, ...] = segmentations
+        elif all([isinstance(seg, da.Image) for seg in segmentations]):
+            segmentation_arrays = tuple([seg.img for seg in segmentations])
+        else:
+            raise Exception(
+                "Segmentation types are not allowed. They should"
+                "all be the same, and either arrays, og daria images."
```
Comment (jwboth) [inline src/daria/analysis/segmentationcomparison.py:197]:
```suggestion
                "all be the same, and either arrays, or daria images."
```

**Comment by jwboth** [comment]
> Do you want to check over the unresolved conversations before I merge?

Looks all good to me. Feel free to resolve all conversations and merge.
